### PR TITLE
[SPARK-47367][PYTHON][CONNECT] Support Python data sources with Spark Connect

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?

This PR adds support for registering and using Python data sources with Spark Connect.

### Why are the changes needed?

To make Python data source work with Spark Connect.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

New unit tests.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: python/pyspark/sql/tests/connect/test_parity_python_datasource.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from pyspark.sql.tests.test_python_datasource import BasePythonDataSourceTestsMixin
+from pyspark.testing.connectutils import ReusedConnectTestCase
+
+class PythonDataSourceParityTests(BasePythonDataSourceTestsMixin, ReusedConnectTestCase):
+    ...
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/connect/test_parity_python_datasource.py:21]:
```suggestion
    pass
```

File: python/pyspark/sql/connect/session.py [hunk 2]
```diff
@@ -743,6 +743,14 @@ def udtf(self) -> "UDTFRegistration":
 
     udtf.__doc__ = PySparkSession.udtf.__doc__
 
+    @property
+    def dataSource(self) -> "DataSourceRegistration":
+        from pyspark.sql.connect.datasource import DataSourceRegistration
+
+        return DataSourceRegistration(self)
+
+    dataSource.__doc__ = PySparkSession.dataSource.__doc__
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/session.py:752]: Let's add 

```
        Notes
        -----
        Supports Spark Connect.
```

into `spark.dataSource` docstring

**Review by HyukjinKwon — APPROVED** [review]
LGTM pending CI

File: python/pyspark/sql/connect/plan.py [hunk 3]
```diff
@@ -2408,6 +2408,46 @@ def __repr__(self) -> str:
...
+        )
+        return plan
+
+    def to_data_source_proto(
+        self, session: "SparkConnectClient"
+    ) -> "proto.CommonInlineUserDefinedDataSource":
+        plan = proto.CommonInlineUserDefinedDataSource()
+        plan.name = self._name
+        plan.python_data_source.CopyFrom(
+            cast(proto.PythonDataSource, self._data_source.to_plan(session))  # type: ignore[arg-type]
```
Comment (ueshin) [inline python/pyspark/sql/connect/plan.py:2446]: Do we need `cast` here?

File: python/pyspark/sql/tests/connect/test_parity_python_datasource.py
Comment (ueshin) [inline python/pyspark/sql/tests/connect/test_parity_python_datasource.py:1]: This module needs to be listed in `dev/sparktestsupport/modules.py`.

**Comment by xinrong-meng** [comment]
Please run `dev/reformat-python`, thanks!

LGTM after resolving the above comments

**Review by dongjoon-hyun — COMMENTED** [review]
Gentle ping, @allisonwang-db .

File: python/pyspark/sql/connect/session.py (hunk 2 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/connect/session.py:752]: I already handled this.

File: python/pyspark/sql/connect/plan.py (hunk 3 above)
Comment (allisonwang-db) [reply python/pyspark/sql/connect/plan.py:2446]: Good catch I think we don't need this cast here.

**Review by ueshin — APPROVED** [review]
LGTM, pending tests.

**Comment by HyukjinKwon** [comment]
Merged to master.

File: python/pyspark/sql/tests/test_python_datasource.py [hunk 4]
```diff
@@ -236,12 +236,10 @@ def reader(self, schema) -> "DataSourceReader":
 
         self.spark.dataSource.register(InMemoryDataSource)
         df = self.spark.read.format("memory").load()
-        self.assertEqual(df.rdd.getNumPartitions(), 3)
```
Comment (zhengruifeng) [inline python/pyspark/sql/tests/test_python_datasource.py:239]: nit, if we want to validate number of partitions in this case, we might use `spark_partition_id`

Comment (HyukjinKwon) [reply python/pyspark/sql/tests/test_python_datasource.py:239]: Made a PR: https://github.com/apache/spark/pull/45720

**Pull request closed without merging**